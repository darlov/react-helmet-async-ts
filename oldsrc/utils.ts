import {TAG_NAMES, TAG_PROPERTIES, ATTRIBUTE_NAMES, SEO_PRIORITY_TAGS} from './constants';
import {
  HelmetProps,
  BaseProps,
  AttributeNamesMap,
  TagPropsType,
  IStateType,
  MetaProps,
  LinkProps,
  ScriptProps
} from "./types";

enum HELMET_PROPS {
  DEFAULT_TITLE = 'defaultTitle',
  DEFER = 'defer',
  ENCODE_SPECIAL_CHARACTERS = 'encodeSpecialCharacters',
  ON_CHANGE_CLIENT_STATE = 'onChangeClientState',
  TITLE_TEMPLATE = 'titleTemplate',
  PRIORITIZE_SEO_TAGS = 'prioritizeSeoTags'
}

const getInnermostProperty = <T extends keyof HelmetProps>(propsList: HelmetProps[], property: T) => {
  for (let i = propsList.length - 1; i >= 0; i -= 1) {
    const props = propsList[i];

    if (Object.hasOwn(props, property)) {
      return props[property];
    }
  }

  return undefined;
};

const getTitleFromPropsList = (propsList: HelmetProps[]) => {
  let innermostTitle = getInnermostProperty(propsList, TAG_NAMES.TITLE);
  const innermostTemplate = getInnermostProperty(propsList, HELMET_PROPS.TITLE_TEMPLATE);
  if (Array.isArray(innermostTitle)) {
    innermostTitle = innermostTitle.join('');
  }
  if (innermostTemplate && innermostTitle) {
    // use function arg to avoid need to escape $ characters
    return innermostTemplate.replace(/%s/g, innermostTitle);
  }

  const innermostDefaultTitle = getInnermostProperty(propsList, HELMET_PROPS.DEFAULT_TITLE);

  return innermostTitle || innermostDefaultTitle || undefined;
};

const getOnChangeClientState = (propsList: HelmetProps[]) =>
  getInnermostProperty(propsList, HELMET_PROPS.ON_CHANGE_CLIENT_STATE) || (() => {
  });

const getAttributesFromPropsList = <T extends ATTRIBUTE_NAMES>(tagType: T, propsList: HelmetProps[]) =>
  propsList
    .map(props => props[tagType])
    .filter((tag): tag is AttributeNamesMap[T] => !!tag)
    .reduce<AttributeNamesMap[T]>((tagAttrs, current) => ({...tagAttrs, ...current}), {});

const getBaseTagFromPropsList = (primaryAttributes: TAG_PROPERTIES[], propsList: HelmetProps[]) =>
  propsList
    .map(props => props.base)
    .filter((tag): tag is BaseProps => !!tag)
    .reverse()
    .reduce<BaseProps[]>((innermostBaseTag, tag) => {
      if (!innermostBaseTag.length) {
        const keys = Object.keys(tag);

        for (let i = 0; i < keys.length; i += 1) {
          const attributeKey = keys[i];
          const lowerCaseAttributeKey = attributeKey.toLowerCase();

          if (primaryAttributes.some(m => m === lowerCaseAttributeKey) && tag[lowerCaseAttributeKey as keyof BaseProps]) {
            return innermostBaseTag.concat(tag);
          }
        }
      }

      return innermostBaseTag;
    }, []);

// eslint-disable-next-line no-console
const warn = (msg: string) => console && typeof console.warn === 'function' && console.warn(msg);

type Flatten<T> = T extends any[] ? T[number] : T;

const getTagsFromPropsList = <T extends keyof TagPropsType,
  TValidTags extends Required<TagPropsType>,
  TTags = TagPropsType[T],
  TElement = Flatten<TTags>>(tagName: T, primaryAttributes: TAG_PROPERTIES[], propsList: TagPropsType[]) => {
  // Calculate list of tags, giving priority innermost component (end of the propslist)
  const approvedSeenTags: { [key in keyof TElement]: any; } = {} as TElement;

  return propsList
    .filter((props): props is TValidTags => {
      if (Array.isArray(props[tagName])) {
        return true;
      }
      if (typeof props[tagName] !== 'undefined') {
        warn(
          `Helmet: ${ tagName } should be of type "Array". Instead found type "${ typeof props[
            tagName
            ] }"`
        );
      }
      return false;
    })
    .map((props) => props[tagName])
    .reverse()
    .reduce((approvedTags, instanceTags) => {
      const instanceSeenTags: { [key in keyof TElement]: any; } = {} as TElement;
      
      instanceTags
        .filter(tag => {
          let primaryAttributeKey: string | undefined = undefined;
          const keys = Object.keys(tag);
          for (let i = 0; i < keys.length; i += 1) {
            const attributeKey = keys[i];
            const lowerCaseAttributeKey = attributeKey.toLowerCase();

            // Special rule with link tags, since rel and href are both primary tags, rel takes priority
            if (
              primaryAttributes.some(m => m === lowerCaseAttributeKey) &&
              !(
                primaryAttributeKey === TAG_PROPERTIES.REL &&
                tag[primaryAttributeKey as keyof typeof tag].toLowerCase() === 'canonical'
              ) &&
              !(
                lowerCaseAttributeKey === TAG_PROPERTIES.REL &&
                tag[lowerCaseAttributeKey as keyof typeof tag].toLowerCase() === 'stylesheet'
              )
            ) {
              primaryAttributeKey = lowerCaseAttributeKey;
            }
            // Special case for innerHTML which doesn't work lowercased
            if (
              primaryAttributes.some(m => m === attributeKey) &&
              (attributeKey === TAG_PROPERTIES.INNER_HTML ||
                attributeKey === TAG_PROPERTIES.CSS_TEXT ||
                attributeKey === TAG_PROPERTIES.ITEM_PROP)
            ) {
              primaryAttributeKey = attributeKey;
            }
          }

          if (!primaryAttributeKey || !tag[primaryAttributeKey as keyof typeof tag]) {
            return false;
          }

          const value: string = tag[primaryAttributeKey as keyof typeof tag].toLowerCase();

          if (!approvedSeenTags[primaryAttributeKey as keyof TElement]) {
            approvedSeenTags[primaryAttributeKey as keyof TElement] = {};
          }

          if (!instanceSeenTags[primaryAttributeKey as keyof TElement]) {
            instanceSeenTags[primaryAttributeKey as keyof TElement] = {};
          }

          if (!approvedSeenTags[primaryAttributeKey as keyof TElement][value]) {
            instanceSeenTags[primaryAttributeKey as keyof TElement][value] = true;
            return true;
          }

          return false;
        }, [])
        .reverse()
        .forEach(tag => approvedTags.push(tag as TElement));

      // Update seen tags with tags from this instance
      const keys = Object.keys(instanceSeenTags);
      for (let i = 0; i < keys.length; i += 1) {
        const attributeKey = keys[i];
        const tagUnion = {
          ...approvedSeenTags[attributeKey as keyof TElement],
          ...instanceSeenTags[attributeKey as keyof TElement],
        };

        approvedSeenTags[attributeKey as keyof TElement] = tagUnion;
      }

      return approvedTags;
    }, [] as TElement[])
    .reverse();
};

const getAnyTrueFromPropsList = <T extends keyof HelmetProps>(propsList: HelmetProps[], checkedTag: T) => {
  if (Array.isArray(propsList) && propsList.length) {
    for (let index = 0; index < propsList.length; index += 1) {
      const prop = propsList[index];
      if (prop[checkedTag]) {
        return true;
      }
    }
  }
  return false;
};
const reducePropsToState = (propsList: HelmetProps[]): IStateType => ({
  baseTag: getBaseTagFromPropsList([TAG_PROPERTIES.HREF], propsList),
  bodyAttributes: getAttributesFromPropsList(ATTRIBUTE_NAMES.BODY, propsList),
  defer: getInnermostProperty(propsList, HELMET_PROPS.DEFER),
  encode: getInnermostProperty(propsList, HELMET_PROPS.ENCODE_SPECIAL_CHARACTERS),
  htmlAttributes: getAttributesFromPropsList(ATTRIBUTE_NAMES.HTML, propsList),
  linkTags: getTagsFromPropsList(
    TAG_NAMES.LINK,
    [TAG_PROPERTIES.REL, TAG_PROPERTIES.HREF],
    propsList
  ),
  metaTags: getTagsFromPropsList(
    TAG_NAMES.META,
    [
      TAG_PROPERTIES.NAME,
      TAG_PROPERTIES.CHARSET,
      TAG_PROPERTIES.HTTPEQUIV,
      TAG_PROPERTIES.PROPERTY,
      TAG_PROPERTIES.ITEM_PROP,
    ],
    propsList
  ),
  noscriptTags: getTagsFromPropsList(TAG_NAMES.NOSCRIPT, [TAG_PROPERTIES.INNER_HTML], propsList),
  onChangeClientState: getOnChangeClientState(propsList),
  scriptTags: getTagsFromPropsList(
    TAG_NAMES.SCRIPT,
    [TAG_PROPERTIES.SRC, TAG_PROPERTIES.INNER_HTML],
    propsList
  ),
  styleTags: getTagsFromPropsList(TAG_NAMES.STYLE, [TAG_PROPERTIES.CSS_TEXT], propsList),
  title: getTitleFromPropsList(propsList),
  titleAttributes: getAttributesFromPropsList(ATTRIBUTE_NAMES.TITLE, propsList),
  prioritizeSeoTags: getAnyTrueFromPropsList(propsList, HELMET_PROPS.PRIORITIZE_SEO_TAGS),
});

export const flattenArray = (possibleArray: any) =>
  Array.isArray(possibleArray) ? possibleArray.join('') : possibleArray;

export { reducePropsToState };

const checkIfPropsMatch = (props: any, toMatch: any) => {
  const keys = Object.keys(props);
  for (let i = 0; i < keys.length; i += 1) {
    // e.g. if rel exists in the list of allowed props [amphtml, alternate, etc]
    if (toMatch[keys[i]] && toMatch[keys[i]].includes(props[keys[i]])) {
      return true;
    }
  }
  return false;
};

export type Prioritizer =
  { element: MetaProps, propsToMatch: typeof SEO_PRIORITY_TAGS["meta"] }
  | { element: LinkProps, propsToMatch: typeof SEO_PRIORITY_TAGS["link"] }
  | { element: ScriptProps, propsToMatch: typeof SEO_PRIORITY_TAGS["script"] };

export const prioritizer = <TElement extends Prioritizer["element"], TMatch extends Prioritizer["propsToMatch"]>(
  elements: TElement[],
  propsToMatch: TMatch) => {

  const result: { priority: TElement[], default: TElement[] } = {priority: [], default: []};

  if (Array.isArray(elements)) {

    for (const element of elements) {
      if (checkIfPropsMatch(element, propsToMatch)) {
        result.priority.push(element);
      } else {
        result.default.push(element);
      }
    }
  }
  return  result;
};

export const without = (obj: any, key: any) => {
  return {
    ...obj,
    [key]: undefined,
  };
}
