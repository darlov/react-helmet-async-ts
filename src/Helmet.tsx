import {FC, memo, ReactNode, useMemo} from "react";
import {IHelmetInstanceState} from "./types";
import {useHelmetContext} from "./HelmetProvider";
import {Title} from "./tags";
import {createActionsData, HelmetScopedContext, HelmetScopedContextData} from "./HelmetScopedProvider";
import {useServerSideEffect} from "./hooks/useServerSideEffect";

interface IHelmetProps {
  defaultTitle?: ReactNode,
  children?: ReactNode
  prioritizeSeoTags?: boolean
}

let counter = 0;
export const Helmet = memo<IHelmetProps>(({children, defaultTitle}) => {
  const rootContext = useHelmetContext();
  const instanceState = useMemo<IHelmetInstanceState>(() => {
    return {id: counter++}
  }, [])

  useServerSideEffect(() => {
      rootContext.addInstance(instanceState);
      return () => rootContext.removeInstance(instanceState);
    },
    () => !rootContext.canUseDOM,
    [instanceState, rootContext.addInstance, rootContext.removeInstance]);

  const context = useMemo<HelmetScopedContextData>(() => {
    return {
      actions: createActionsData(instanceState, rootContext)
    }
  }, [instanceState, rootContext.addItem, rootContext.removeItem])

  return (
    <HelmetScopedContext.Provider value={context}>
      {defaultTitle && <Title>{defaultTitle}</Title>}
      {children}
    </HelmetScopedContext.Provider>);
});

Helmet.displayName = "Helmet";