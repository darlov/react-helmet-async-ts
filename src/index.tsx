export { HelmetContextProvider, useHelmetContext } from "./HelmetProvider"
export { Helmet } from "./Helmet"
export * from "./tags"
export { HELMET_ATTRIBUTE } from "./Types"