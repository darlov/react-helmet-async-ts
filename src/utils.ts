import { Dispatch, SetStateAction } from "react";
import {
  ArrayElement,
  DOCUMENT_TITLE_INSTANCE_ID,
  IHelmetInstanceState,
  IHelmetState,
  IHelmetTags,
  LinkProps,
  MetaProps,
  primaryLinkAttributes,
  primaryMetaAttributes,
} from "./types";

export const addUniqueItem = <T, K extends T[] | undefined>(
  items: K,
  item: T,
  keySelector?: (item: T) => any
): K => {
  if (!items) {
    return [item] as K;
  }

  const exist =
    keySelector !== undefined
      ? items.some(m => keySelector(m) === keySelector(item))
      : items.some(m => m === item);
  return exist ? items : ([...items, item] as K);
};

export const removeAction =
  <T, K extends T[] | undefined>(
    action: Dispatch<SetStateAction<K>>,
    keySelector?: (item: T) => any
  ) =>
  (item: T) => {
    action(items => {
      return keySelector !== undefined
        ? (items?.filter(m => keySelector(m) !== keySelector(item)) as K)
        : (items?.filter(m => m !== item) as K);
    });
  };
export const addAction = <T, K extends T[] | undefined>(
  action: Dispatch<SetStateAction<K>>,
  keySelector?: (item: T) => any
) => {
  return (item: T) => action(items => addUniqueItem<T, K>(items, item, keySelector));
};

export namespace _ {
  export type TSelector<T> = ((elem: T) => any) | keyof T;

  export const sortBy = <T, K extends T[] | undefined>(array: K, selector: TSelector<T>): K => {
    if(array === undefined){
      return array;
    }
    
    if (typeof selector === "function") {
      return array.sort((x, y) =>
        selector(x) > selector(y) ? 1 : selector(x) < selector(y) ? -1 : 0
      ) as K;
    }

    return array.sort((x, y) =>
      x[selector] > y[selector] ? 1 : x[selector] < y[selector] ? -1 : 0
    ) as K;
  };

  export const clear = <T>(array: T[] | undefined) => {
    if (array !== undefined && array.length > 0) {
      array.splice(0, array.length);
    }
  };

  export const isEmpty = (obj: Record<string, any>) => {
    return Object.keys(obj).length === 0;
  };

  type TArray = any[] | undefined;
  export const isEmptyArray = (...obj: TArray[]) => {
    return obj.every(m => m === undefined || m.length === 0);
  };

  type InferKey<T> = T extends Partial<Record<infer K, any>> ? K : never;
  type InferValue<T> = T extends Partial<Record<any, infer V>> ? V : never;

  export const toPairs = <T extends Partial<Record<string, any>>>(obj: T) => {
    return Object.entries(obj) as [InferKey<T>, InferValue<T>][];
  };

  type TValSelReturn<
    TItem,
    T extends ((arg: TItem, index: number) => any) | undefined
  > = T extends (arg: TItem, index: number) => infer K ? K : T extends undefined ? TItem : never;

  type TValueSelector<T> = ((item: T, index: number) => T) | ((item: T, index: number) => any);

  export const groupBy = <
    T,
    K extends keyof any,
    V extends TValSelReturn<T, TFunc>,
    TFunc extends TValueSelector<T>
  >(
    items: T[],
    keySelector: (item: T, index: number) => K | undefined,
    valueSelector?: TFunc
  ): Map<K, TValSelReturn<T, TFunc>[]> => {
    return items.reduce((result, item, index) => {
      const key = keySelector(item, index);
      if (key !== undefined) {
        if (!result.has(key)) {
          result.set(key, []);
        }
        result.get(key)!.push(valueSelector && valueSelector(item, index));
        return result;
      }
      return result;
    }, new Map<K, TValSelReturn<T, TFunc>[]>());
  };
}

const findMetaPrimaryAttribute = (meta: MetaProps) => {
  const foundAttr = primaryMetaAttributes.find(attr => meta[attr] !== undefined);

  if (foundAttr !== undefined) {
    return `${foundAttr}_${meta[foundAttr]}`;
  }

  return undefined;
};

const findLinkPrimaryAttribute = (link: LinkProps) => {
  const foundAttr = primaryLinkAttributes.find(attr => {
    return link[attr] !== undefined ? !(attr === "rel" && link[attr] === "stylesheet") : false;
  });

  if (foundAttr !== undefined) {
    return `${foundAttr}_${link[foundAttr]}`;
  }

  return undefined;
};

const mergeAllToOne = <T extends keyof IHelmetTags, TElement extends ArrayElement<IHelmetTags[T]>>(
  key: T,
  result: TElement | undefined,
  instance: IHelmetInstanceState,
  emptyStateFallback?: (result: TElement | undefined) => TElement | undefined
): TElement | undefined => {
  if (instance.emptyState) {
    if (emptyStateFallback !== undefined) {
      return emptyStateFallback(result);
    }

    return undefined;
  }

  const values = instance[key] as TElement[] | undefined;
  return (values || []).reduce((prev, current) => {
    return { ...current };
  }, result);
};

const mergeAllToAll = <T extends keyof IHelmetTags, TElement extends ArrayElement<IHelmetTags[T]>>(
  key: T,
  result: TElement[],
  instance: IHelmetInstanceState
): TElement[] => {

  if (instance.emptyState) {
    _.clear(result);
  }
  
  if (instance[key]) {
      const values = instance[key] as TElement[];
      result.push(...values);
  }

  return result;
};

const mergeAllByPrimaryAttribute = <T extends keyof IHelmetTags, TElement = IHelmetTags[T]>(
  tagProp: T,
  result: TElement[],
  instance: IHelmetInstanceState,
  primaryAttributeSelector: (tag: TElement) => string | undefined
) => {
  if (instance.emptyState) {
    _.clear(result);
  }else {
    const instanceInputTags = instance[tagProp] as TElement[] | undefined;
    if (instanceInputTags) {
      if (result.length === 0) {
        result.push(...instanceInputTags);
      } else {
        const instanceGrouped = _.groupBy(
            instanceInputTags,
            primaryAttributeSelector,
            (item, index) => [item, index] as [TElement, number]
        );
        const resultGrouped = _.groupBy(
            result,
            primaryAttributeSelector,
            (item, index) => [item, index] as [TElement, number]
        );

        for (const [attr, instanceTags] of instanceGrouped.entries()) {
          const resultTags = resultGrouped.get(attr);

          if (resultTags === undefined) {
            result.push(...instanceTags.map(([tags]) => tags));
          } else if (instanceTags.length !== resultTags.length) {
            result = result.filter(m => !resultTags.some(([tag]) => m === tag));
            result.push(...instanceTags.map(([tag]) => tag));
          } else {
            for (let i = 0; i < instanceTags.length; i++) {
              const [instanceTag] = instanceTags[i];
              const [resultTag, resultIndex] = resultTags[i];

              result[resultIndex] = {...resultTag, ...instanceTag};
            }
          }
        }
      }
    }
  }

  return result;
};

export const buildState = (instances: IHelmetInstanceState[]): IHelmetState => {
  const state: IHelmetState = {
    titleTag: undefined,
    baseTag: undefined,
    bodyTag: undefined,
    htmlTag: undefined,
    metaTags: [],
    styleTags: [],
    scriptTags: [],
    linkTags: [],
    noscriptTags: [],
  };

  const documentTitleInstance = instances.find(m => m.id == DOCUMENT_TITLE_INSTANCE_ID);

  const titleEmptyStateFallback = (result: IHelmetState["titleTag"]) =>
    documentTitleInstance !== undefined && documentTitleInstance.titleTags !== undefined
      ? documentTitleInstance.titleTags[0]
      : result;

  for (const instance of instances) {
    state.titleTag = mergeAllToOne("titleTags", state.titleTag, instance, titleEmptyStateFallback);
    state.baseTag = mergeAllToOne("baseTags", state.baseTag, instance);
    state.bodyTag = mergeAllToOne("bodyTags", state.bodyTag, instance);
    state.htmlTag = mergeAllToOne("htmlTags", state.htmlTag, instance);
    state.styleTags = mergeAllToAll("styleTags", state.styleTags, instance);
    state.scriptTags = mergeAllToAll("scriptTags", state.scriptTags, instance);
    state.noscriptTags = mergeAllToAll("noscriptTags", state.noscriptTags, instance);
    state.metaTags = mergeAllByPrimaryAttribute(
      "metaTags",
      state.metaTags,
      instance,
      findMetaPrimaryAttribute
    );
    state.linkTags = mergeAllByPrimaryAttribute(
      "linkTags",
      state.linkTags,
      instance,
      findLinkPrimaryAttribute
    );
  }

  return state;
};
