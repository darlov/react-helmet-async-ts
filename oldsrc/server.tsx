import React, {ClassAttributes, createElement, DetailedHTMLProps, HTMLAttributes, ReactElement} from 'react';
import {
  ATTRIBUTE_NAMES,
  HELMET_ATTRIBUTE,
  REACT_TAG_MAP,
  SEO_PRIORITY_TAGS,
  TAG_NAMES,
  TAG_PROPERTIES,
} from './constants';
import {flattenArray, prioritizer} from './utils';
import {
  HelmetDatum,
  HelmetHTMLBodyDatum,
  HelmetHTMLElementDatum,
  HelmetServerState,
  IStateType,
  TitleProps,
  toEntries
} from "./types";

const SELF_CLOSING_TAGS = [TAG_NAMES.NOSCRIPT, TAG_NAMES.SCRIPT, TAG_NAMES.STYLE];


const encodeSpecialCharacters = (str: string, encode = true) => {
  if (!encode) {
    return String(str);
  }

  return String(str)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#x27;');
};

const generateElementAttributesAsString = (data: TypedTagsAttributes) =>
  toEntries(data.attributes).reduce((str, [key, value]) => {
    const attr = value !== undefined
      ? `${key}="${value}"`
      : `${key}`;
    return str ? `${str} ${attr}` : attr;
  }, '');

const generateTitleAsString = (data: TypedTagsAttrTitle, encode?: boolean) => {
  const attributeString = generateElementAttributesAsString(data);
  const flattenedTitle = flattenArray(data.title);
  return attributeString
    ? `<${data.type} ${HELMET_ATTRIBUTE}="true" ${attributeString}>${encodeSpecialCharacters(
      flattenedTitle,
      encode
    )}</${data.type}>`
    : `<${data.type} ${HELMET_ATTRIBUTE}="true">${encodeSpecialCharacters(
      flattenedTitle,
      encode
    )}</${data.type}>`;
};

const generateTagsAsString = (data: Exclude<TagType, TypedTagsAttributes>, encode?: boolean): string =>
  (data.tags as any[]).reduce((str, tag) => {
    const attributeHtml = Object.keys(tag)
      .filter(
        attribute =>
          !(attribute === TAG_PROPERTIES.INNER_HTML || attribute === TAG_PROPERTIES.CSS_TEXT)
      )
      .reduce((string, attribute) => {
        const attr =
          typeof tag[attribute] === 'undefined'
            ? attribute
            : `${attribute}="${encodeSpecialCharacters(tag[attribute], encode)}"`;
        return string ? `${string} ${attr}` : attr;
      }, '');

    const tagContent = tag.innerHTML || tag.cssText || '';

    const isSelfClosing = SELF_CLOSING_TAGS.indexOf(data.type) === -1;

    return `${str}<${data.type} ${HELMET_ATTRIBUTE}="true" ${attributeHtml}${
      isSelfClosing ? `/>` : `>${tagContent}</${data.type}>`
    }`;
  }, '');

type extractGeneric<Type> = Type extends DetailedHTMLProps<HTMLAttributes<infer T>, any> ? T : never

const convertElementAttributesToReactProps = <P extends HTMLAttributes<T>, T extends HTMLElement>(attributes: ClassAttributes<T> & P, initProps: any = {}): HTMLAttributes<T> =>
  toEntries(attributes).reduce((obj, [key, value]) => {
    obj[REACT_TAG_MAP[key as keyof typeof REACT_TAG_MAP] || key] = value;
    return obj;
  }, initProps);

const generateTitleAsReactComponent = (title: string | undefined, attributes: TitleProps) => {
  // assigning into an array to define toString function on it
  const initProps = {
    key: title,
    [HELMET_ATTRIBUTE]: true,
  };
  const props = convertElementAttributesToReactProps(attributes, initProps);

  return [createElement(TAG_NAMES.TITLE, props, title)];
};

const generateTagsAsReactComponent = <T extends Exclude<TagType, TypedTagsAttributes>>(data: T): ReactElement[] => {
  return data.tags.map((tag, i) => {
    const mappedTag: any = {
      key: i,
      [HELMET_ATTRIBUTE]: true,
    };

    toEntries(tag).forEach(([attribute, value]) => {
      if (attribute  as any == TAG_PROPERTIES.INNER_HTML || attribute as any === TAG_PROPERTIES.CSS_TEXT) {
        const content = (tag as any).innerHTML || (tag as any).cssText;
        mappedTag.dangerouslySetInnerHTML = {__html: content};
      } else {
        const mappedAttribute = REACT_TAG_MAP[attribute as unknown as keyof typeof REACT_TAG_MAP] || attribute;
        mappedTag[mappedAttribute] = value;
      }
    });

    return createElement(data.type, mappedTag);
  })
}


type TypedItem<T> = { type: T };
type TypedTagsAttr<T, TKey extends keyof IStateType> = TypedItem<T> & { attributes: IStateType[TKey] };
type TypedTags<T, TKey extends keyof IStateType> = TypedItem<T> & { tags: IStateType[TKey] };

type TypedTagsAttrTitle = TypedTagsAttr<TAG_NAMES.TITLE, "titleAttributes"> & { title: IStateType["title"] };
type TypedTagsAttrBody = TypedTagsAttr<ATTRIBUTE_NAMES.BODY, "bodyAttributes">;
type TypedTagsAttrHtml = TypedTagsAttr<ATTRIBUTE_NAMES.HTML, "htmlAttributes">;

type TypedTagsAttributes = TypedTagsAttrTitle | TypedTagsAttrBody | TypedTagsAttrHtml;

type TagType =
  TypedTagsAttributes
  | TypedTags<TAG_NAMES.STYLE, "styleTags">
  | TypedTags<TAG_NAMES.SCRIPT, "scriptTags">
  | TypedTags<TAG_NAMES.NOSCRIPT, "noscriptTags">
  | TypedTags<TAG_NAMES.META, "metaTags">
  | TypedTags<TAG_NAMES.LINK, "linkTags">
  | TypedTags<TAG_NAMES.BASE, "baseTag">;


type TagOrAttribute<T extends TagType> =
  T extends { type: ATTRIBUTE_NAMES.HTML } ? HelmetHTMLElementDatum :
    T extends { type: ATTRIBUTE_NAMES.BODY } ? HelmetHTMLBodyDatum :
      HelmetDatum;

const getMethodsForTag = <T extends TagType, TReturn extends TagOrAttribute<T>>(data: T, encode?: boolean): TagOrAttribute<T> => {
  switch (data.type) {
    case TAG_NAMES.TITLE:
      return {
        toComponent: () => generateTitleAsReactComponent(data.title, data.attributes),
        toString: () => generateTitleAsString(data, encode),
      } as any;
    case ATTRIBUTE_NAMES.BODY:
      return {
        toComponent: () => convertElementAttributesToReactProps(data.attributes),
        toString: () => generateElementAttributesAsString(data),
      } as any;
    case ATTRIBUTE_NAMES.HTML:
      return {
        toComponent: () => convertElementAttributesToReactProps(data.attributes),
        toString: () => generateElementAttributesAsString(data),
      } as any;
    default:
      return {
        toComponent: () => generateTagsAsReactComponent(data),
        toString: () => generateTagsAsString(data, encode),
      } as any;
  }
};

const getPriorityMethods = ({metaTags, linkTags, scriptTags, encode}: IStateType) => {
  const meta = prioritizer(metaTags, SEO_PRIORITY_TAGS.meta);
  const link = prioritizer(linkTags, SEO_PRIORITY_TAGS.link);
  const script = prioritizer(scriptTags, SEO_PRIORITY_TAGS.script);

  // need to have toComponent() and toString()
  const priorityMethods = {
    toComponent: () => [
      ...generateTagsAsReactComponent({type: TAG_NAMES.META, tags: meta.priority}),
      ...generateTagsAsReactComponent({type: TAG_NAMES.LINK, tags: link.priority}),
      ...generateTagsAsReactComponent({type: TAG_NAMES.SCRIPT, tags: script.priority}),
    ][0],
    toString: () =>
      // generate all the tags as strings and concatenate them
      `${getMethodsForTag({type: TAG_NAMES.META, tags: meta.priority}, encode)} ${getMethodsForTag(
        {type: TAG_NAMES.LINK, tags: link.priority},
        encode
      )} ${getMethodsForTag({type: TAG_NAMES.SCRIPT, tags: script.priority}, encode)}`,
  };

  return {
    priorityMethods,
    metaTags: meta.default,
    linkTags: link.default,
    scriptTags: script.default,
  };
};

const mapStateOnServer = (props: IStateType): HelmetServerState => {
  const {
    baseTag,
    bodyAttributes,
    encode,
    htmlAttributes,
    noscriptTags,
    styleTags,
    title = '',
    titleAttributes,
    prioritizeSeoTags,
  } = props;
  let {linkTags, metaTags, scriptTags} = props;
  let priorityMethods = {
    toComponent: (): ReactElement => (<></>),
    toString: () => '',
  };
  if (prioritizeSeoTags) {
    ({priorityMethods, linkTags, metaTags, scriptTags} = getPriorityMethods(props));
  }
  return {
    priority: priorityMethods,
    base: getMethodsForTag({type: TAG_NAMES.BASE, tags: baseTag}, encode),
    bodyAttributes: getMethodsForTag({type: ATTRIBUTE_NAMES.BODY, attributes: bodyAttributes}, encode),
    htmlAttributes: getMethodsForTag({type: ATTRIBUTE_NAMES.HTML, attributes: htmlAttributes}, encode),
    link: getMethodsForTag({type: TAG_NAMES.LINK, tags: linkTags}, encode),
    meta: getMethodsForTag({type: TAG_NAMES.META, tags: metaTags}, encode),
    noscript: getMethodsForTag({type: TAG_NAMES.NOSCRIPT, tags: noscriptTags}, encode),
    script: getMethodsForTag({type: TAG_NAMES.SCRIPT, tags: scriptTags}, encode),
    style: getMethodsForTag({type: TAG_NAMES.STYLE, tags: styleTags}, encode),
    title: getMethodsForTag({type: TAG_NAMES.TITLE, title, attributes: titleAttributes}, encode),
  };
};

export default mapStateOnServer;
