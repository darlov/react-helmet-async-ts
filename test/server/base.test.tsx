import {renderToStaticMarkup} from 'react-dom/server';
import {Base, Helmet, IHelmetData} from '../../src';
import {render} from './utils';

describe('server', () => {
  describe('Declarative API', () => {
    it('renders base tag as React component', () => {
      const context: IHelmetData = {};
      render(
        <Helmet>
          <Base target="_blank" href="http://localhost/"/>
        </Helmet>,
        context
      );

      expect(context.state).toBeDefined();
      const head = context.state!;
      
      expect(head.base).toBeDefined();
      expect(head.base.toComponent).toBeDefined();

      const baseComponent = head!.base.toComponent();

      expect(baseComponent).toBeDefined()
      expect(baseComponent).toEqual(expect.objectContaining({type: 'base'}));

      const markup = renderToStaticMarkup(baseComponent);

      expect(markup).toMatchSnapshot();
    });

    it('renders base tags as string', () => {
      const context: IHelmetData = {};
      render(
        <Helmet>
          <Base target="_blank" href="http://localhost/"/>
        </Helmet>,
        context
      );

      const head = context.state;

      expect(head).toBeDefined();
      expect(head!.base).toBeDefined();
      expect(head!.base.toString).toBeDefined();
      expect(head!.base.toString()).toMatchSnapshot();
    });
  });
});
