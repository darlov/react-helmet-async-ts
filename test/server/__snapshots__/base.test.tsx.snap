// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`server > Declarative API > renders base tag as React component 1`] = `"<base target="_blank" href="http://localhost/" data-rh="true"/>"`;

exports[`server > Declarative API > renders base tags as string 1`] = `"<base target="_blank" href="http://localhost/" data-rh="true"/>"`;
